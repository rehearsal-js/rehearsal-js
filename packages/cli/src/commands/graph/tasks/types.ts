export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
