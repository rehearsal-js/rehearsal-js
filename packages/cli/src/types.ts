import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
