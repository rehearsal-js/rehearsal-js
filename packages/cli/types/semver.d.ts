declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
