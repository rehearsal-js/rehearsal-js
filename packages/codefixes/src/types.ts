import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
