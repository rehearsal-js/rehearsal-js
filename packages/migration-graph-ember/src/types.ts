/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Package } from '@rehearsal/migration-graph-shared';
import type { EmberAddonPackage } from './entities/ember-addon-package.js';
import type { EmberAppPackage } from './entities/ember-app-package.js';

export type EmberProjectPackage = Package | EmberAddonPackage | EmberAppPackage;
