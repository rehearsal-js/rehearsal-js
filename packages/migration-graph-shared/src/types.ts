import type { Package } from './entities/package.js';

export type UniqueNode = {
  key: string;
  synthetic?: boolean;
};

export type PackageNode = UniqueNode & {
  pkg: Package;
  converted?: boolean;
};

export type ModuleNode = UniqueNode & {
  path: string;
  parsed?: boolean;
};
