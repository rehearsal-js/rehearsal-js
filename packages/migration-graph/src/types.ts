import { Package } from '@rehearsal/migration-graph-shared';

import { Graph } from './utils/graph';

export type UniqueGraphNode = {
  key: string;
};

export type PackageNode = UniqueGraphNode & {
  pkg: Package;
  converted?: boolean;
  modules: Graph<ModuleNode>;
};

export type ModuleNode = UniqueGraphNode & {
  path: string;
  parsed?: boolean;
  meta?: unknown;
};
