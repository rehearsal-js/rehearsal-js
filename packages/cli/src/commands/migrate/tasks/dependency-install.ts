import { resolve } from 'node:path';
import { existsSync, promises as fs } from 'node:fs';
import { ListrTask } from 'listr2';

import { addDep } from '@rehearsal/utils';
import { PackageJson } from 'type-fest';
import { MigrateCommandContext, MigrateCommandOptions } from '../../../types.js';

export const REQUIRED_DEPENDENCIES = [
  '@types/node',
  '@typescript-eslint/eslint-plugin',
  '@typescript-eslint/parser',
  'eslint',
  'eslint-config-prettier',
  'eslint-plugin-prettier',
  'prettier',
  'typescript',
];

// get the name of dependency from those two format:
// - foo
// - foo@{version}
// Be aware that a package name would start with @, e.g @types/node
function extractDepName(dep: string): string {
  const reg = /^(@?[^@]+)/g;
  const matched = dep.match(reg);
  return matched ? matched[0] : dep;
}

// check if package.json has all required dependecies
// from rehearsal default and user config
export async function shouldRunDepInstallTask(
  options: MigrateCommandOptions,
  context: Partial<MigrateCommandContext> = {}
): Promise<boolean> {
  const { basePath } = options;
  // add extra required dependencies from user config if there is any
  let dependencies: string[] = [];
  let devDependencies = REQUIRED_DEPENDENCIES;
  if (context.userConfig?.hasDependencies) {
    // check if package.json has all the dependencies listed in user config
    dependencies = [...dependencies, ...context.userConfig.dependencies];
    devDependencies = [...devDependencies, ...context.userConfig.devDependencies];
  }

  const packageJsonPath = resolve(basePath, 'package.json');
  if (existsSync(packageJsonPath)) {
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    for (const d of dependencies) {
      if (!packageJson.dependencies || !packageJson.dependencies[extractDepName(d)]) {
        return true;
      }
    }

    for (const d of devDependencies) {
      if (!packageJson.devDependencies || !packageJson.devDependencies[extractDepName(d)]) {
        return true;
      }
    }
  } else {
    // no package.json
    return true;
  }
  return false;
}

export function depInstallTask(
  options: MigrateCommandOptions,
  context?: Partial<MigrateCommandContext>
): ListrTask {
  return {
    title: 'Install dependencies',
    enabled: (): boolean => !options.dryRun,
    skip: async (ctx: MigrateCommandContext): Promise<boolean> =>
      !(await shouldRunDepInstallTask(options, ctx)),
    task: async (ctx: MigrateCommandContext, task): Promise<void> => {
      // If context is provide via external parameter, merge with existed
      if (context) {
        ctx = { ...ctx, ...context };
      }

      let dependencies: string[] = [];
      let devDependencies = REQUIRED_DEPENDENCIES;
      // install custom dependencies
      if (ctx.userConfig?.hasDependencies) {
        task.output = `Install dependencies from config`;
        dependencies = [...dependencies, ...ctx.userConfig.dependencies];
        devDependencies = [...devDependencies, ...ctx.userConfig.devDependencies];
        if (ctx.userConfig?.hasPostInstallHook) {
          task.output = `Run postInstall hook from config`;
          await ctx.userConfig.postInstall();
        }
      }

      const errorMessages: string[] = [];

      // for deps
      if (dependencies.length) {
        try {
          task.output = `Installing dependecies: ${dependencies.join(', ')}`;
          await addDep(dependencies, false, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(dependencies, false));
        }
      }

      // for devDeps
      if (devDependencies.length) {
        try {
          task.output = `Installing devDependencies: ${devDependencies.join(', ')}`;
          await addDep(devDependencies, true, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(devDependencies, true));
        }

        if (errorMessages.length) {
          throw new Error(errorMessages.join('\n'));
        }
      }
    },
    // will print and keep what dpe is currently installing at bottom bar
    options: { persistentOutput: true, bottomBar: Infinity },
  };
}

function formatErrorMessage(deps: string[], dev: boolean): string {
  const depType = dev ? 'devDependencies' : 'dependencies';
  return [
    `We ran into an error when installing ${depType}, please install the following as ${depType} and try again.`,
    `${deps.map((d) => `  - ${d}`).join('\n')}`,
  ].join('\n');
}
