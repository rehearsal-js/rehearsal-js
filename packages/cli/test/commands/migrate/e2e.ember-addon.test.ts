import { resolve } from 'node:path';
import { readFileSync, readdirSync, promises as fs, existsSync } from 'node:fs';
import { describe, test, expect, beforeAll, afterAll } from 'vitest';
import { getEmberAddonProject, getEmberAddonGjsProject } from '@rehearsal/test-support';
import { readJSONSync } from 'fs-extra/esm';
import { TSConfig } from '../../../src/types.js';
import { REQUIRED_DEPENDENCIES } from '../../../src/commands/migrate/tasks/dependency-install.js';
import { runBin } from '../../test-helpers/index.js';
import type { Project } from 'fixturify-project';
import type { PackageJson } from 'type-fest';

describe('migrate - ember addon: e2e', () => {
  describe('ember addon', () => {
    let project: Project;

    beforeAll(async () => {
      project = getEmberAddonProject();
      await project.write();

      await runBin('migrate', ['--ci'], {
        cwd: project.baseDir,
      });
    });

    afterAll(() => {
      project.dispose();
    });

    test('should have the correct files in the base directory', async () => {
      expectFilesExist(project.baseDir, [
        'ember-cli-build.js',
        'testem.js',
        '.ember-cli',
        'index.js',
      ]);
      await verifyPackageJson(project.baseDir);
      verifyReport(project.baseDir);
      verifyTsConfig(project.baseDir);
    });

    test('should have the correct files in the app components directory', () => {
      const appComponentsPath = resolve(project.baseDir, 'app', 'components');
      expectFilesExist(appComponentsPath, ['greet.js']);
    });

    test('should have the correct files in the addon components directory', () => {
      const addonComponentsPath = resolve(project.baseDir, 'addon', 'components');
      expectFilesExist(addonComponentsPath, ['greet.ts', 'greet.hbs']);

      expect(
        readFileSync(resolve(addonComponentsPath, 'greet.ts'), { encoding: 'utf-8' })
      ).toMatchSnapshot();
      expect(
        readFileSync(resolve(addonComponentsPath, 'greet.hbs'), { encoding: 'utf-8' })
      ).toMatchSnapshot();
    });

    test('should have the correct files in the config directory', () => {
      const configPath = resolve(project.baseDir, 'config');
      expectFilesExist(configPath, ['ember-try.js', 'environment.js']);
    });

    test('should have the correct files in the tests directory', () => {
      const testsPath = resolve(project.baseDir, 'tests');
      expectFilesExist(testsPath, ['test-helper.ts', 'index.html']);

      let testHelperContent = readFileSync(resolve(testsPath, 'test-helper.ts'), {
        encoding: 'utf-8',
      });
      testHelperContent = testHelperContent.replace(
        new RegExp(`${project.baseDir}`, 'g'),
        '<tmp-path>'
      );
      expect(testHelperContent).toMatchSnapshot();
    });

    test('should have the correct files in the tests acceptance directory', () => {
      const testsAcceptancePath = resolve(project.baseDir, 'tests', 'acceptance');
      expectFilesExist(testsAcceptancePath, ['addon-template-test.ts']);

      let addonTemplateTestContent = readFileSync(
        resolve(testsAcceptancePath, 'addon-template-test.ts'),
        {
          encoding: 'utf-8',
        }
      );
      addonTemplateTestContent = addonTemplateTestContent.replace(
        new RegExp(`${project.baseDir}`, 'g'),
        '<tmp-path>'
      );
      expect(addonTemplateTestContent).toMatchSnapshot();
    });

    test('should have the correct files in the tests dummy app directory', () => {
      const testsDummyAppPath = resolve(project.baseDir, 'tests', 'dummy', 'app');
      expectFilesExist(testsDummyAppPath, ['app.ts', 'router.ts', 'index.html']);

      let testsDummyAppContent = readFileSync(resolve(testsDummyAppPath, 'app.ts'), {
        encoding: 'utf-8',
      });
      testsDummyAppContent = testsDummyAppContent.replace(
        new RegExp(`${project.baseDir}`, 'g'),
        '<tmp-path>'
      );
      expect(testsDummyAppContent).toMatchSnapshot();

      let testsDummyRouterContent = readFileSync(resolve(testsDummyAppPath, 'router.ts'), {
        encoding: 'utf-8',
      });
      testsDummyRouterContent = testsDummyRouterContent.replace(
        new RegExp(`${project.baseDir}`, 'g'),
        '<tmp-path>'
      );
      expect(testsDummyRouterContent).toMatchSnapshot();
    });

    test('should have the correct files in the tests dummy app templates directory', () => {
      const testsDummyTemplatesPath = resolve(
        project.baseDir,
        'tests',
        'dummy',
        'app',
        'templates'
      );
      expectFilesExist(testsDummyTemplatesPath, ['application.hbs', 'index.hbs']);

      expect(
        readFileSync(resolve(testsDummyTemplatesPath, 'application.hbs'), { encoding: 'utf-8' })
      ).toMatchSnapshot();
      expect(
        readFileSync(resolve(testsDummyTemplatesPath, 'index.hbs'), { encoding: 'utf-8' })
      ).toMatchSnapshot();
    });
  });

  describe('ember addon gjs', () => {
    let project: Project;

    beforeAll(async () => {
      project = getEmberAddonGjsProject();
      await project.write();

      await runBin('migrate', ['--ci'], {
        cwd: project.baseDir,
      });
    });

    afterAll(() => {
      project.dispose();
    });

    test('should have the correct files in the base directory', async () => {
      expectFilesExist(project.baseDir, [
        'ember-cli-build.js',
        'testem.js',
        '.ember-cli',
        'index.js',
      ]);
      await verifyPackageJson(project.baseDir);
      verifyReport(project.baseDir);
      verifyTsConfig(project.baseDir);
    });

    test('should have the correct files in the config directory', () => {
      const configPath = resolve(project.baseDir, 'config');
      expectFilesExist(configPath, ['ember-try.js', 'environment.js']);
    });

    test('should have the correct files in the tests directory', () => {
      const testsPath = resolve(project.baseDir, 'tests');
      expectFilesExist(testsPath, ['test-helper.ts', 'index.html']);

      let testHelperContent = readFileSync(resolve(testsPath, 'test-helper.ts'), {
        encoding: 'utf-8',
      });
      testHelperContent = testHelperContent.replace(
        new RegExp(`${project.baseDir}`, 'g'),
        '<tmp-path>'
      );
      expect(testHelperContent).toMatchSnapshot();
    });

    test('should have the correct files in the tests dummy app directory', () => {
      const testsDummyAppPath = resolve(project.baseDir, 'tests', 'dummy', 'app');
      expectFilesExist(testsDummyAppPath, ['app.ts', 'router.ts', 'index.html']);

      let appContent = readFileSync(resolve(testsDummyAppPath, 'app.ts'), { encoding: 'utf-8' });
      appContent = appContent.replace(new RegExp(`${project.baseDir}`, 'g'), '<tmp-path>');
      let routerContent = readFileSync(resolve(testsDummyAppPath, 'router.ts'), {
        encoding: 'utf-8',
      });
      routerContent = routerContent.replace(new RegExp(`${project.baseDir}`, 'g'), '<tmp-path>');
      expect(appContent).toMatchSnapshot();
      expect(routerContent).toMatchSnapshot();
    });

    test('should have the correct files in the tests dummy app templates directory', () => {
      const testsDummyTemplatesPath = resolve(
        project.baseDir,
        'tests',
        'dummy',
        'app',
        'templates'
      );
      expectFilesExist(testsDummyTemplatesPath, ['application.hbs']);

      expect(
        readFileSync(resolve(testsDummyTemplatesPath, 'application.hbs'), { encoding: 'utf-8' })
      ).toMatchSnapshot();
    });
  });
});

function expectFilesExist(path: string, filesArray: string[]): void {
  const fileList = readdirSync(path);

  for (const file of filesArray) {
    expect(fileList).toContain(file);
  }
}

async function verifyPackageJson(basePath: string): Promise<void> {
  const packageJson = JSON.parse(
    await fs.readFile(resolve(basePath, 'package.json'), 'utf-8')
  ) as PackageJson;
  expect(packageJson?.scripts?.['lint:tsc']).toBe('tsc --noEmit');

  const devDeps = packageJson.devDependencies;
  for (const requiredDep of REQUIRED_DEPENDENCIES) {
    expect(Object.keys(devDeps || {})).toContain(requiredDep);
  }
}

function verifyReport(basePath: string): void {
  const reportPath = resolve(basePath, '.rehearsal');
  const sarifPath = resolve(reportPath, 'migrate-report.sarif');
  const jsonPath = resolve(reportPath, 'migrate-report.json');
  expect(existsSync(reportPath)).toBeTruthy();
  expect(existsSync(sarifPath)).toBeTruthy();
  expect(existsSync(jsonPath)).toBeTruthy();
}

function verifyTsConfig(basePath: string): void {
  const tsConfig = readJSONSync(resolve(basePath, 'tsconfig.json')) as TSConfig;
  expect(tsConfig).matchSnapshot();
}
