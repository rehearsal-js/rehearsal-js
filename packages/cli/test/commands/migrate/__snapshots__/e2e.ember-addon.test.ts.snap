// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the addon components directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@glimmer/component'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Component from \\"@glimmer/component\\";

export default class Greet extends Component {
  get name() {
    return \\"Sue\\";
  }
}
"
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the addon components directory 2`] = `"Hello {{this.name}}, from an addon!"`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the base directory 1`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
  "include": [
    "addon/components/greet.ts",
    "tests/acceptance/addon-template-test.ts",
    "tests/dummy/app/app.ts",
    "tests/dummy/app/router.ts",
    "tests/test-helper.ts",
  ],
}
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests acceptance directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { module, test } from \\"qunit\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/test-helpers'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { visit, currentURL } from \\"@ember/test-helpers\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { setupApplicationTest } from \\"ember-qunit\\";

/* @ts-expect-error @rehearsal TODO TS7006: Parameter 'hooks' implicitly has an 'any' type. */
module(\\"Acceptance | addon-template\\", function (hooks) {
  setupApplicationTest(hooks);

  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'assert' implicitly has an 'any' type. */
  test(\\"visiting /\\", async function (assert) {
    await visit(\\"/\\");

    assert.equal(currentURL(), \\"/\\");

    assert
      .dom(\\"#app-container\\")
      .hasText(\\"Hello Sue, from an addon!\\", \\"The user can see Hello World\\");
  });
});
"
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'dummy/app'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Application from \\"dummy/app\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import * as QUnit from \\"qunit\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/test-helpers'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { setApplication } from \\"@ember/test-helpers\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'qunit-dom'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { setup } from \\"qunit-dom\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { start } from \\"ember-qunit\\";

setApplication(Application.create(config.APP));

setup(QUnit.assert);

start();
"
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests dummy app directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/application'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Application from \\"@ember/application\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-resolver'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Resolver from \\"ember-resolver\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-load-initializers'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import loadInitializers from \\"ember-load-initializers\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";

export default class App extends Application {
  modulePrefix = config.modulePrefix;
  podModulePrefix = config.podModulePrefix;
  Resolver = Resolver;
}

loadInitializers(App, config.modulePrefix);
"
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests dummy app directory 2`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/routing/router'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import EmberRouter from \\"@ember/routing/router\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";

export default class Router extends EmberRouter {
  location = config.locationType;
  rootURL = config.rootURL;
}

/* @ts-expect-error @rehearsal TODO TS2339: Property 'map' does not exist on type 'typeof Router'. */
Router.map(function () {});
"
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests dummy app templates directory 1`] = `
"
              <div id='app-container'>
                {{outlet}}
              </div>
            "
`;

exports[`migrate - ember addon: e2e > ember addon > should have the correct files in the tests dummy app templates directory 2`] = `"<Greet/>"`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the addon components directory 1`] = `
"
          {{! @glint-expect-error @rehearsal TODO TS2792: Cannot find module '@glint/environment-ember-template-imports/-private/dsl'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? }}
          {{! @glint-expect-error @rehearsal TODO TS7006: Parameter 'ùö™' implicitly has an 'any' type. }}
          const Flea = <template>
            <p>Hello, {{@name}}!</p>
          </template>;
          
          {{! @glint-expect-error @rehearsal TODO TS2792: Cannot find module '@glint/environment-ember-template-imports/-private/dsl'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? }}
          {{! @glint-expect-error @rehearsal TODO TS7006: Parameter 'ùö™' implicitly has an 'any' type. }}
          <template>
            <Flea @name=\\"Littlest\\"/>
          </template>
        "
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the addon components directory 2`] = `
"
          /* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'addon-template/components/branch'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
          import Branch from 'addon-template/components/branch'; 

          {{! @glint-expect-error @rehearsal TODO TS2792: Cannot find module '@glint/environment-ember-template-imports/-private/dsl'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? }}
          {{! @glint-expect-error @rehearsal TODO TS7006: Parameter 'ùö™' implicitly has an 'any' type. }}
          <template>
            <Branch/>
          </template>
        "
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the base directory 1`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
  "include": [
    "addon/components/leaf.gts",
    "addon/components/branch.gts",
    "addon/components/trunk.gts",
    "tests/dummy/app/app.ts",
    "tests/dummy/app/router.ts",
    "tests/test-helper.ts",
  ],
}
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the tests directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'dummy/app'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Application from \\"dummy/app\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import * as QUnit from \\"qunit\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/test-helpers'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { setApplication } from \\"@ember/test-helpers\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'qunit-dom'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { setup } from \\"qunit-dom\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-qunit'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import { start } from \\"ember-qunit\\";

setApplication(Application.create(config.APP));

setup(QUnit.assert);

start();
"
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the tests dummy app directory 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/application'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Application from \\"@ember/application\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-resolver'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import Resolver from \\"ember-resolver\\";
/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module 'ember-load-initializers'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import loadInitializers from \\"ember-load-initializers\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";

export default class App extends Application {
  modulePrefix = config.modulePrefix;
  podModulePrefix = config.podModulePrefix;
  Resolver = Resolver;
}

loadInitializers(App, config.modulePrefix);
"
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the tests dummy app directory 2`] = `
"/* @ts-expect-error @rehearsal TODO TS2792: Cannot find module '@ember/routing/router'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option? */
import EmberRouter from \\"@ember/routing/router\\";
/* @ts-expect-error @rehearsal TODO TS7016: Could not find a declaration file for module 'dummy/config/environment'. '<tmp-path>/tests/dummy/config/environment.js' implicitly has an 'any' type. */
import config from \\"dummy/config/environment\\";

export default class Router extends EmberRouter {
  location = config.locationType;
  rootURL = config.rootURL;
}

/* @ts-expect-error @rehearsal TODO TS2339: Property 'map' does not exist on type 'typeof Router'. */
Router.map(function () {});
"
`;

exports[`migrate - ember addon: e2e > ember addon gjs > should have the correct files in the tests dummy app templates directory 1`] = `
"{{page-title \\"Dummy\\"}}

{{!-- The following component displays Ember's default welcome message. --}}
<WelcomePage />
{{!-- Feel free to remove this! --}}

{{outlet}}"
`;
